### PR TITLE
only enable query key scaling during fp16

### PR DESCRIPTION
**Pull request by gshennvm**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

Enable query key scaling only during fp16, since this is when TE enables query key scaling. https://github.com/NVIDIA/TransformerEngine/blob/666539f36275fa9c0fbc99f9ea50f2d6e29e336f/transformer_engine/pytorch/attention.py#L940

### CODE REVIEW

**Comment by ericharper** [comment]
jenkins

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 1]
```diff
@@ -1544,6 +1544,11 @@ def build_transformer_config(self) -> TransformerConfig:
 
         attention_softmax_in_fp32 = False  # not currently used in NeMo unless apply_query_key_layer_scaling is True
         apply_query_key_layer_scaling = self.cfg.get('apply_query_key_layer_scaling', False)
+
+        if apply_query_key_layer_scaling and not model_parallel_config.fp16:
+            logging.warning("apply_query_key_layer_scaling is only enabled when using FP16, setting it to False")
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1549]: I don't think model_parallel_config.fp16 is the right check though. That arg is for fp16 + megatron_amp_O2.

Maybe we should just check trainer.precision?

Comment (gshennvm) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1549]: can I check against `self.torch_dtype`? https://github.com/NVIDIA/NeMo/blob/8b27f3a68879a78d35fd7d5bac72fb792cf18da8/nemo/collections/nlp/models/language_modeling/megatron_base_model.py#L112

**Comment by gshennvm** [comment]
jenkins

**Comment by gshennvm** [comment]
jenkins

**Comment by gshennvm** [comment]
jenkins

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -1544,6 +1544,15 @@ def build_transformer_config(self) -> TransformerConfig:
 
         attention_softmax_in_fp32 = False  # not currently used in NeMo unless apply_query_key_layer_scaling is True
         apply_query_key_layer_scaling = self.cfg.get('apply_query_key_layer_scaling', False)
+
+        fp16_enabled = self.trainer.precision in [16, '16', '16-mixed']
+        if apply_query_key_layer_scaling:
+            if fp16_enabled:
+                os.environ["NVTE_APPLY_QK_LAYER_SCALING"] = "1"
+            else:
+                logging.warning("apply_query_key_layer_scaling is only enabled when using FP16, setting it to False")
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1553]: Should we set the env var here as well to "0" ?

Comment (gshennvm) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py]: good point, otherwise it will error 366cf61f82ada5e89f5af5c45eed598309ec6875

**Comment by gshennvm** [comment]
jenkins

**Pull request merged**